Sync API example (#120)
JupiterOne/jupiterone-client-nodejs: examples/sync-api/README.md

# TL;DR

> If you want to form a relationship with a `_key`, you must include the `_source`

Reviewer: ```suggestion
If you want to form a relationship using a `_key`, you must include the `_source`
```

and `_scope` of the entity that already exists in the graph.

> Assuming that the entity you are uploading is the `to` and the entity that you
> are forming a relationship already exists in the graph and therefore is the
> `from`:

Reviewer: ```suggestion
Assuming that the entity you are uploading is the `to` and the entity that you
are forming a relationship with already exists in the graph and therefore is the
`from`:
```

This will _NOT_ work:

```
{
    _fromEntityKey: entityFrom.entity._key,
    _toEntityKey: entityTo.entity._key,
}
```

This will work:

```
{
    _fromEntitySource: entityFrom.entity._source,
    _fromEntityScope: entityFrom.entity._integrationInstanceId,
    _fromEntityKey: entityFrom.entity._key,
    _toEntityKey: entityTo.entity._key,
}
```

> # Creating Relationship Between Entities You Own With Entities You Do Not

Reviewer: ```suggestion
# Creating Relationships Between Entities You Own And Entities You Do Not
```

Our goal here is to create relationships from one entity to another that you

> might not own. Ownership is a concept in JupiterOne

Reviewer: Was there supposed to be more to this sentence? 🤔

Author: lmao yes :sadge: I wanted to release it. I'll fix.

> While this appears trivial, without knowing the details around how to form
> relationships, you might wonder why your data doesn't appear in the graph.

Reviewer: ```suggestion
This may seem trivial. But if you don't know exactly how relationships are formed in J1, you might not produce the graph you expect.
```

Side note, this sentence is depressing because it is true. :sadge:

AND from = 'testing'
THAT USES << CodeRepo
```

and see our entities in the graph with the expected relationships formed.

This experiment outlines a common use-case:

You want to add new data to the JupiterOne graph and you want to form
relationships with that data.

What makes this use-case stand out is that you do not have an ID for your entity
yet.

This means to form a relationship with one sync job you've got to utilize the
`_key` property of your entity! The catch is that you must be specific in how
you do so.

Let's dig in.

## Acquiring Data in the Graph to Form Our Relationships

Assuming you already have your data prepared and ready to send to J1, the next
thing to do is to gather the entities that already exist in the JupiterOne graph
so that you can form relationships with them.

Writing our query is straightforward enough:

```
FIND github_repo
    WITH displayName = ('graph-veracode' OR 'graph-knowbe4' OR 'graph-azure' OR 'graph-wazuh' OR 'graph-enrichment-examples' OR 'graph-whois' OR 'graph-zeit')
```

Sweet! We have `CodeRepos` that we're going to form relationships with.

> Here's an example payload:

Reviewer: ```suggestion
Here's an example query response payload:
```

Remember the comment from earlier: we do not have the `CodeModule` `_id` yet!
This is important because these two options are _NOT_ equal in how they behave.
If you were to form a relationship between the `_id` of the `CodeRepo` with the
`_key` of the CodeModule, this will work. This works because the `_id` is unique
amongst all of the data in your account. The `_key` value is _NOT_ globally
unique (i.e. two entities can have the same `_key`).

When you form a relationship with two `_key` values and you do not specify the
`source` and the `scope` of the data that already exists in the graph, the
JupiterOne software does not understand what entity you're talking about and
ultimately doesn't create the relationship. Since two entities could have the
same `_key`, the software needs more information in order to identify the entity
you're referencing.

### How do you get more information?

Use the `source` and `scope` of your JupiterOne data alongside the `_key`!

```
{
    _fromEntitySource: string;
    _toEntitySource: string;
    _fromEntityScope: string;
    _toEntityScope: string;
}
```

This:

- `CodeRepo` `_key` -> `CodeModule` `_key`
- `CodeRepo` `_id` -> `CodeModule` `_key`

> Becomes this:

Reviewer: ```suggestion
Must actually be this:
```
